File: sdk/src/main/scala/iie/udps/common/hcatalog/scala/SerHCatOutputFormat.scala
package iie.udps.common.hcatalog.scala

import org.apache.hadoop.mapreduce.{ OutputFormat, Job, JobContext, TaskAttemptContext, InputSplit, RecordWriter }
import org.apache.hadoop.io.NullWritable

import org.apache.hadoop.conf.Configuration
import org.apache.hive.hcatalog.mapreduce.{ HCatOutputFormat, HCatBaseOutputFormat }
import org.apache.hive.hcatalog.mapreduce.OutputJobInfo
import org.apache.hive.hcatalog.data.HCatRecord
import org.apache.hive.hcatalog.data.schema.HCatSchema

import org.apache.spark.SerializableWritable

class SerHCatOutputFormat extends OutputFormat[NullWritable, SerializableWritable[HCatRecord]] {

  val instance = new HCatOutputFormat();

  def getRecordWriter(
    context: TaskAttemptContext) = {
    context.getConfiguration().set("mapred.task.id",
      context.getTaskAttemptID().toString())
    context.getConfiguration().setInt("mapred.task.partition",
      context.getTaskAttemptID().getTaskID().getId())
    val writer = instance
      .getRecordWriter(context)
    new RecordWriter[NullWritable, SerializableWritable[HCatRecord]]() {

      def write(key: NullWritable, value: SerializableWritable[HCatRecord]) {
        writer.write(key, value.value)
      }

      def close(context: TaskAttemptContext) {
        writer.close(context)
      }

    };
  }

  def checkOutputSpecs(context: JobContext) =
    instance.checkOutputSpecs(context)

  def getOutputCommitter(context: TaskAttemptContext) =
    instance.getOutputCommitter(context)

}

object SerHCatOutputFormat {
  def setOutput(outputJob: Job, outputJobInfo: OutputJobInfo) =
    HCatOutputFormat.setOutput(outputJob, outputJobInfo)

  def getTableSchema(conf: Configuration) =
    HCatBaseOutputFormat.getTableSchema(conf)

  def setSchema(outputJob: Job, schema: HCatSchema) =
    HCatOutputFormat.setSchema(outputJob, schema)

}

File: sdk/src/main/java/iie/common/hcatalog/SerHCatOutputFormat.java
package iie.common.hcatalog;

import java.io.IOException;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.io.WritableComparable;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.JobContext;
import org.apache.hadoop.mapreduce.OutputCommitter;
import org.apache.hadoop.mapreduce.OutputFormat;
import org.apache.hadoop.mapreduce.RecordWriter;
import org.apache.hadoop.mapreduce.TaskAttemptContext;
import org.apache.hive.hcatalog.data.HCatRecord;
import org.apache.hive.hcatalog.data.schema.HCatSchema;
import org.apache.hive.hcatalog.mapreduce.HCatOutputFormat;
import org.apache.hive.hcatalog.mapreduce.OutputJobInfo;
import org.apache.spark.SerializableWritable;

public class SerHCatOutputFormat extends
		OutputFormat<WritableComparable, SerializableWritable> {

	private HCatOutputFormat instance = new HCatOutputFormat();

	@Override
	public RecordWriter<WritableComparable, SerializableWritable> getRecordWriter(
			TaskAttemptContext context) throws IOException,
			InterruptedException {
		context.getConfiguration().set("mapred.task.id",
				context.getTaskAttemptID().toString());
		context.getConfiguration().setInt("mapred.task.partition",
				context.getTaskAttemptID().getTaskID().getId());
		final RecordWriter<WritableComparable<?>, HCatRecord> writer = instance
				.getRecordWriter(context);
		return new RecordWriter<WritableComparable, SerializableWritable>() {

			@Override
			public void write(WritableComparable key, SerializableWritable value)
					throws IOException, InterruptedException {
				writer.write(key, (HCatRecord) value.value());
			}

			@Override
			public void close(TaskAttemptContext context) throws IOException,
					InterruptedException {
				writer.close(context);
			}

		};
	}

	@Override
	public void checkOutputSpecs(JobContext context) throws IOException,
			InterruptedException {
		instance.checkOutputSpecs(context);
	}

	@Override
	public OutputCommitter getOutputCommitter(TaskAttemptContext context)
			throws IOException, InterruptedException {
		return instance.getOutputCommitter(context);
	}

	public static void setOutput(Job outputJob, OutputJobInfo outputJobInfo)
			throws IOException {
		HCatOutputFormat.setOutput(outputJob, outputJobInfo);
	}

	public static HCatSchema getTableSchema(Configuration conf)
			throws IOException {
		return HCatOutputFormat.getTableSchema(conf);
	}

	public static void setSchema(Job outputJob, HCatSchema schema)
			throws IOException {
		HCatOutputFormat.setSchema(outputJob, schema);
	}

}
